proposed/MPID Messaging.md

```diff
@@ -0,0 +1,295 @@
+    ::crypto::hash::sha512::hash(signed_header)
+}
+pub fn mpid_message_name(mpid_message: &MpidMessage) -> ::routing::NameType {
+    mpid_header_name(mpid_message.signed_header)
+}
+
+pub fn sign_mpid_header(mpid_header: &MpidHeader, private_key: PrivateKey) -> Vec<u8> {
+    ::sodiumoxide::crypto::sign::sign(encode(mpid_header), private_key)
+}
+pub fn parse_signed_header(signed_header: &Vec<u8>, public_key: PublicKey) -> Result<MpidHeader, ::routing::error::Error> {
+    decode::<MpidHeader>(::sodiumoxide::crypto::sign::verify(signed_header, public_key))
+}
+```
+
+## Outbox
+
+This is a simple data structure for now and will be a hash map of serialised and encrypted `MpidMessage`s.  There will be one such map per MPID (owner), held on the MpidManagers, and synchronised by them at churn events.
+
+## Inbox
+
+This is an even simpler structure and again there will be one per MPID (owner), held on the MpidManagers, and synchronised by them at churn events.
+
+This can be implemented as a `Vec<(sender: ::routing::NameType, signed_header: Vec<u8>)>`.
+
+## Messaging Format Among Nodes
+
+Messages between Clients and MpidManagers will utilise [`::routing::structured_data::StructuredData`][5]:
+```rust
+let sd_for_mpid_message = StructuredData {
+    type_tag: MPID_MESSAGE_TAG,
+    identifier: mpid_message_name(mpid_message),
+    data: encode(mpid_message),
+    previous_owner_keys: vec![],
+    version: 0,
+    current_owner_keys: vec![mpid_message.sender],
+    previous_owner_signatures: vec![]
+}
+```
+```rust
+let sd_for_mpid_header = StructuredData {
+    type_tag: MPID_HEADER_TAG,
+    identifier: mpid_message_name(mpid_message),  // or mpid_header_name(signed_header)
+    data: mpid_message.signed_header,  // or inbox.signed_header
+    previous_owner_keys: vec![],
+    version: 0,
+    current_owner_keys: vec![mpid_message.sender],  // or vec![inbox.sender]
+    previous_owner_signatures: vec![]
+}
+```
+
+## Message Flow
+
+```
+                    MpidManagers(A)    MpidManagers(B)
+                   /      *                   *        \
+MpidClient(A) ->  -       *                   *         -  <- MpidClient(B)
+                   \      *                   *        /
+
+```
+1. `MpidClient(A)` sends an `MpidMessage` to `MpidManagers(A)` as a Put request.
+1. `MpidManagers(A)` store this message in their outbox for A and send the `signed_header` component to `MpidManagers(B)` as a Put request, unless the outbox is full in which case a PutFailure response is returned to `MpidClient(A)`.
+1. `MpidManagers(B)` try to store the `(sender, signed_header)` in their inbox for B.  If successful, they forward it to `MpidClient(B)` as a Put request immediately or as soon as it appears online.  If unsuccessful (e.g. the inbox is full or sender has been blacklisted), they reply to `MpidManagers(A)` with a PutFailure, who then remove the message from the outbox and send a PutFailure response to `MpidClient(A)`.
```

Comment:
> easiest way to handle this initially is to change the type of MessageHeader::sender from ::routing::NameType to ::sodiumoxide::crypto::sign::PublicKey

Yes, I too think that would be better.

> This will do away with the need for Clients and Managers to hold a copy of MPIDs

This however I am unable to understand. Let me put it in detail so that you can explain better.

If the Clients have no idea of the sender's MPID's then what is the use of having a signed message ? One answer by @maqi was:
> the signed stuff here is to ensure a security between a messaging peer (on vault level).
messages will be stored in OutBox and Inbox, once it is restored and forwarded again, the content is open to be modified and routing sign can not eliminate that risk

if a message can be modified that way then one can well argue that entire message along with the Key replaced such that the whole combination `Vec<(sender: Mpid, signed_header: Vec<u8>)>` is valid. I thought in this case the Client would check `sender` against the copy of `MPID`s that he has and if it matches, he can be certain the message has not been tampered with and if it does not then he it depends if he trusts the unknown `sender`. If Clients are assumed to *not* hold MPID copies then what purpose does signing solve ?